SSE 도입하여 API 추가

roomie00 opened the pull request:
### 📌 PR 타입(하나 이상의 PR 타입을 선택해주세요)
- 기능 추가
- 리팩토링

### 📌 관련 이슈
#183 

### ✨ 반영 브랜치
feat/183 -> develop

### 📝 변경 사항
- [x] RabbitMQ consumer 추가 (fanout용)
- [x] SSE 도입
    - [x] 운행 일지
    - [x] 차량 상세보기 내 실시간 위치
    - [x] 대시보드 내 실시간 위치

### ➕ 추가 메모
- power off 시에도 SSE 알림이 와야 대시보드에서 제대로 정보를 제공할 수 있을 것 같습니다. (이는 추가 이슈를 팔 예정)
- 여러 변경으로 인해 PR 사이즈가 다소 큽니다...
- Open-in-view 설정을 false로 수정했습니다. SSE Http 연결이 유지되다 보니 DB 커넥션풀을 점유하고 있는 문제가 발생해서 수정했습니다. 작성되어 있는 테스트코드들은 잘 돌아가는 것 확인했으나, 이와 관련해서 문제가 발생할 수도 있습니다. (점유하는 CP가 늘어나는 문제는 해결했습니다. -> 직접적인 DB 및 entity 접근 없도록 코드 작성)
- 기존 polling 방식으로 사용했던 API 들은 남겨두었습니다.
    -> 비교할 수도 있을 것 같고, 추후 대시보드까지(power-off까지) 된다면 기존 API 삭제하겠습니다.

### 🐛 테스트 결과 (테스트 결과가 있다면 넣어주세요)

roomie00 committed:
feat: #183 fanout Exchange 설정 추가

roomie00 committed:
fix: #183 LocalDateTime.now()에 타임존 seoul로 추가

roomie00 committed:
refactor: #183 CurrentGpsLog -> CoordinatesInfo 로 클래스명 변경

roomie00 committed:
feat: #183 SSE 도입
- 운행일지 내 GPS 로그 청킹 전송
- 차량 상세보기 내 실시간 GPS 스트리밍
- 업체 대시보드 내 차량들 현 위치 스트리밍

roomie00 committed:
Merge branch 'develop' of https://github.com/Kernel360/KBE5-Thisway-BE into feat/183

roomie00 committed:
fix: #183 open-in-view 설정 false로 수정

roomie00 committed:
docs: #183 rabbitMQConfig 내 주석 삭제

ClszK commented on src/main/java/org/thisway/triplog/service/StreamCoordinatesService.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    @Async
+    public void sendChunkedTripLog(SseEmitter emitter, List<CoordinatesInfo> allTripLogs) {
+        List<List<CoordinatesInfo>> chunkedData = chunk(allTripLogs, 60);
```
60을 상수화 시키는 것도 좋을꺼 같습니다.

Jeongyounghyeon replied:
값이 유동적으로 변경될 수 있게 할 수 있을까요?
주기가 달라질 수 있는 상황을 고려하여 하드 코딩보다 변수가 되도 좋을 것 같습니다!

roomie00 replied:
해당 부분은 이미 있는 데이터를 청킹해서 보내는 것으로 실시간과는 관련이 없습니다. 따라서 변경될 일이 극히 드물다고 판단하여 하드코딩해뒀습니다. 상수화 시키는 정도로 수정하는 것은 괜찮은데, 이 부분은 주기처럼 쿼리파라미터 등으로 설정값을 받을 필요가 없다고 생각합니다.

SeoYoonP commented on src/main/java/org/thisway/component/streaming/SseConnection.java:
```diff
@@ -0,0 +1,84 @@
+        final SseEmitter emitter;
+        final LocalDateTime connectionTime;
+        final AtomicBoolean initialChunkCompleted = new AtomicBoolean(false);
+        final Queue<Object> bufferedLiveData = new ConcurrentLinkedQueue<>();
```
ConcurrentLinkedQueue에 크기 제한이 없을때 실시간 데이터가 계속 들어온다면 무제한으로 데이터가 쌓이지 않을까 싶습니다. 
`new LinkedBlockingQueue<>(1000)` 이런식으로 크기 제한을 해두면 어떨까 합니다.
더해서 버퍼가 가득찼을때는
```
if (context.bufferedLiveData.size() >= MAX_BUFFER_SIZE) {
    context.bufferedLiveData.poll(); // 가장 오래된 항목 제거
}
context.bufferedLiveData.add(data);
```
이렇게 오래된 데이터를 제거하고 새 데이터를 추가하는 방식으로 하면 좀 더 안전하지 않을까 합니다!

roomie00 replied:
이 부분의 존재 이유는 차량 상세보기 페이지에서 클라이언트가 예를 들어 이미 1시간 정도 운행한 차량에 대해 조회할 때 이전 1시간의 데이터를 청킹해서 다 보내기 전에 실시간으로 들어오는 데이터를 임시 보관하는 곳으로, 모든 운행 중의 실시간 GPS 정보가 이곳에 담기는 것은 아닙니다.

DB에서 가져오는 지나간 데이터가 다 청킹으로 보내졌을 시 initialChunkCompleted가 true로 바뀌게 되며, 이때부터는 큐에 담지 않고 바로 send 합니다. 따라서 1000건 이상의 데이터가 쌓이는 경우는 매우 드물 것 같습니다. 그리고 현재 운행 정보 내의 모든 GPS 데이터를 보여주는 것이 저희 정책이다보니 오래된 데이터를 제거하는 것은 이와도 맞지 않을 것 같습니다.

roomie00 replied:
우려하시는 부분에 공감하여 sseEmitter timeout 설정을 통해 이를 담고 있는 map 내의 메타데이터도 사라지게 수정하겠습니다.

SeoYoonP commented on src/main/java/org/thisway/triplog/controller/TripLogController.java:
```diff
@@ -39,7 +50,15 @@ public ResponseEntity<CurrentTripLogResponse> getCurrentGpsLogs(
     }
 
-    @GetMapping()
+    @GetMapping(value = "/current/stream/{id}", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
+    public SseEmitter getVehicleCurrentGpsLogsSseEmitter(@PathVariable Long id, @RequestParam("token") String token) {
+        if (!jwtTokenUtil.isValid(token)) {
+            throw new CustomException(ErrorCode.AUTH_UNAUTHENTICATED);
+        }
+        return streamCoordinatesService.createStreamForVehicle(id, jwtTokenUtil.getUsernameFromToken(token));
+    }
+
+    @GetMapping
```
현재 토큰 검증이 컨트롤러 메서드 내에서 처리되고 있는것 같은데 컨트롤러 계층에서 처리되는건 적합하지 않아보여요!

ClszK replied:
이 부분은 Spring Security에서 AccessToken을 Header로 옮기는 작업을 진행하면 해결될 부분일꺼에요

roomie00 replied:
이거는 전에 서윤님 안계셨을 때 공유했던 내용이었는데 SSE 특성 상 Http 연결을 지속해야하기 때문에 클라이언트에서 EventSource 객체를 사용해야합니다. 해당 객체는 request header를 커스텀할 수 없어 쿼리 파라미터로 작성했습니다.

이를 해결하기 위한 방법은 쿠키를 사용하는 방법과 쿼리 파라미터인데, 쿠키를 사용하기에는 security 관련 부분을 전체적으로 수정해야해서 일단은 쿼리 파라미터로 두겠습니다.

SeoYoonP commented on src/main/java/org/thisway/triplog/controller/TripLogController.java:
```diff
@@ -53,4 +72,13 @@ public ResponseEntity<TripLogDetailResponse> getTripLogDetail(@PathVariable Long
         return ResponseEntity.status(HttpStatus.OK)
                 .body(tripLogService.getTripLogDetails(id));
     }
+
+    @GetMapping(value = "/detail/stream/{id}", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
+    public SseEmitter getTripLogsStream(@PathVariable Long id, @RequestParam("token") String token) {
+
+        if (!jwtTokenUtil.isValid(token)) {
+            throw new CustomException(ErrorCode.AUTH_UNAUTHENTICATED);
+        }
+        return streamCoordinatesService.createStreamForTripLog(id);
+    }
```
+53 to +61 라인과 같은 의견입니다!

roomie00 replied:
위에 참고 부탁드립니다.

SeoYoonP reviewed (commented):
전체적으로 너무 잘 구성해 주셨고 너무 고생하신것같습니다. 컨트롤러에서 토큰검증 로직처리하는 부분은 리팩토링으로 넘겨도 될 것 같고 SseConnection에 ConcurrentLinkedQueue 부분만 확인부탁드려요!

Jeongyounghyeon commented on src/main/java/org/thisway/component/streaming/SseConnection.java:
```diff
@@ -0,0 +1,84 @@
+    private final Map<String, SseContext> emitters = new ConcurrentHashMap<>();
+
+    public SseEmitter createSseEmitter(String key) {
+        SseEmitter sseEmitter = new SseEmitter(0L);
```
적절한 timeout 값이 있어야할 것 같습니다!

roomie00 replied:
적절한 timeout의 값을 어떻게 정의해야할지 모르겠어서 우선 무한으로 뒀습니다. 한 30분 정도로 설정해서 수정하겠습니다.

Jeongyounghyeon commented on src/main/java/org/thisway/component/streaming/SseEventSender.java:
```diff
@@ -0,0 +1,45 @@
+    public void sendToPrefix(String prefix, String eventName, Object data) {
+        Set<String> keys = sseConnection.findKeysByPrefix(prefix);
+
+        for (String key : keys) {
+            sendLiveDataWithBuffering(key, eventName, data);
+        }
```
반복문 도중 하나의 클라이언트에게 전송 실패했을 경우, 다음 모든 클라이언트에 대해서 전송을 하지 않고 있는 것 같습니다!
예외처리 후 다음 클라이언트에게도 전송 시도해도 좋을 것 같습니다!

roomie00 replied:
넵 수정하도록 하겠습니다. 👍

Jeongyounghyeon commented on src/main/java/org/thisway/log/dto/request/gpsLog/GpsLogEntry.java:
```diff
@@ -11,4 +11,11 @@ public record GpsLogEntry(
         String sum,
         String bat
 ) {
+    public String getSec() {
```
따로 추가하신 이유가 있으실까요?

roomie00 replied:
실시간 대시보드에서 가장 최근의 데이터 (마지막)를 받아와야해서 추가했습니다.
물론 저희 애뮬레이터에서 잘 정렬해서 오긴 하지만 혹시나를 가정해서 추가했습니다.

@Getter 를 안쓴 이유는.... 딱히 없습니다.... ㅎㅎ

Jeongyounghyeon replied:
record를 사용하고 있어서 기본적으로 제공해주는 getter를 사용하지 않는 이유를 여쭤봤습니다!

Jeongyounghyeon commented on src/main/java/org/thisway/triplog/dto/CoordinatesInfo.java:
```diff
@@ -0,0 +1,29 @@
+        );
+    }
+
+    public static CoordinatesInfo fromGpsEntry (GpsLogEntry gpsLogEntry, Long vehicleId) {
+        return new CoordinatesInfo(
+                vehicleId,
+                Double.parseDouble(gpsLogEntry.lat()) / 1_000_000.0,
+                Double.parseDouble(gpsLogEntry.lon()) / 1_000_000.0,
+                Integer.parseInt(gpsLogEntry.ang())
+        );
+    }
```
두 팩토리 메서드 중 해당 메서드만 이름이 구체적인 이유가 있을까요?

roomie00 replied:
그러면 다 from으로 바꿔두겠습니다.

Jeongyounghyeon commented on src/main/java/org/thisway/triplog/service/StreamCoordinatesService.java:
```diff
@@ -0,0 +1,144 @@
+    private final SseEventSender sseEventSender;
+
+    public SseEmitter createStreamForTripLog(Long tripId) {
+        SseEmitter emitter = new SseEmitter(0L);
```
이부분만 SseConnection에 있는 createEmiter를 사용하지 않은 이유가 있으실까요?

roomie00 replied:
이 부분은 청킹해서 데이터를 보내는 부분으로 모두 보내면 해당 로직 안에서 complete를 호출하기 때문에 map 안에 sse 연결에 대한 메타 데이터를 저장할 필요가 없으며, 오히려 적절하지 않다고 판단하여 다음과 같이 해뒀습니다.

이 또한 timeout 설정은 10분으로 해두겠습니다.

roomie00 committed:
fix: #183 리뷰 반영 수정
- record내 메서드명 from으로 수정
- SSE Emitter 타임아웃 설정 (각 10분, 30분)
- gpsLog broadcast 시 에러 예외 처리
- chunk_size 상수화

Jeongyounghyeon reviewed (approved):
👍

roomie00 committed:
Merge branch 'develop' of https://github.com/Kernel360/KBE5-Thisway-BE into feat/183

# Conflicts:
#	src/main/java/org/thisway/log/repository/LogRepository.java

roomie00 committed:
Merge branch 'develop' of https://github.com/Kernel360/KBE5-Thisway-BE into feat/183

roomie00 committed:
fix: # 183 develop 병합
